clients/go/dkv/client.go

```diff
@@ -0,0 +1,248 @@
+package dkv
+
+import (
+	"fmt"
+	"github.com/OneOfOne/xxhash"
+	"github.com/dgraph-io/ristretto"
+	"github.com/flipkart-incubator/dkv/pkg/ctl"
+	"github.com/flipkart-incubator/dkv/pkg/serverpb"
+
+	"log"
+)
+
+type DKVServerRole string
+
+const (
+	noRole     DKVServerRole = "UNKNOWN"
+	masterRole               = "MASTER"
+	slaveRole                = "SLAVE"
+)
+
+func GetNodeTypeByReadConsistency(rc serverpb.ReadConsistency) DKVServerRole {
+	if rc == serverpb.ReadConsistency_LINEARIZABLE {
+		return masterRole
+	}
+	return slaveRole
+}
+
+type DKVNode struct {
+	Host string `json:"host"`
+	Port int    `json:"port"`
+}
+
+type DKVNodeSet struct {
+	Name  string    `json:"name"`
+	Nodes []DKVNode `json:"nodes"`
+}
+
+type DKVShard struct {
+	Name     string                        `json:"name"`
+	Topology map[DKVServerRole]*DKVNodeSet `json:"topology"`
+}
+
+func (s DKVShard) GetNodesByType(nodeType ...DKVServerRole) (*DKVNodeSet, error) {
+	for _, v := range nodeType {
+		if val, ok := s.Topology[v]; ok {
+			return val, nil
+		}
+	}
+	return nil, fmt.Errorf("valid DKV node type must be given")
+}
+
+type ShardProvider interface {
+	ProvideShard(key []byte) (*DKVShard, error)
+	ProvideShards(keys ...[]byte) (map[*DKVShard][][]byte, error)
+}
+
+type KeyHashBasedShardProvider struct {
+	ShardConfiguration []DKVShard
+}
+
+func (p KeyHashBasedShardProvider) getShardID(key []byte) int {
+	h := xxhash.New64()
+	h.Write(key)
+	hash := h.Sum64()
+	var id = (hash & 0xFFFF) % uint64(len(p.ShardConfiguration))
+	//LongHashFunction xx = LongHashFunction.xx();
+	//long hsh = xx.hashBytes(key);
+	//return (int) ((hsh & 0xFFFF) % shardConfiguration.getNumShards());
+	return int(id)
+}
+func (p KeyHashBasedShardProvider) ProvideShard(key []byte) (*DKVShard, error) {
+	shardId := p.getShardID(key)
+	return &p.ShardConfiguration[shardId], nil
+}
+
+func (p KeyHashBasedShardProvider) ProvideShards(keys ...[]byte) (map[*DKVShard][][]byte, error) {
```

Comment:
Please include GoDoc for this public API

Reply:
Done